File: 04-bffs/REGENERA-04-BFFS-10-10/packages/shared/src/audit-chain.mjs
import { createHash } from 'node:crypto';
import { canonicalJson } from './canonical-json.mjs';

function hash(value) {
  return createHash('sha256').update(value).digest('hex');
}

export class AuditChain {
  #events = [];

  append(event) {
    const previousHash = this.#events.at(-1)?.hash ?? '0'.repeat(64);
    const payload = { sequence: this.#events.length + 1, previousHash, event: structuredClone(event) };
    const entry = { ...payload, hash: hash(canonicalJson(payload)) };
    this.#events.push(entry);
    return structuredClone(entry);
  }

  entries() {
    return structuredClone(this.#events);
  }

  static verify(entries) {
    let previousHash = '0'.repeat(64);
    for (let index = 0; index < entries.length; index += 1) {
      const entry = entries[index];
      const payload = { sequence: index + 1, previousHash, event: entry.event };
      if (entry.sequence !== index + 1) return false;
      if (entry.previousHash !== previousHash) return false;
      if (entry.hash !== hash(canonicalJson(payload))) return false;
      previousHash = entry.hash;
    }
    return true;
  }
}

File: 05-core-bank/REGENERA-05-CORE-BANKING-10-10/src/main/kotlin/com/regenera/core/AuditChain.kt
package com.regenera.core

import java.time.Instant

data class AuditEvent(
    val sequence: Long,
    val eventType: String,
    val actorId: String,
    val subjectId: String,
    val payload: Map<String, String>,
    val occurredAt: Instant,
    val previousHash: String,
    val eventHash: String,
)

class AuditChain {
    private val events = mutableListOf<AuditEvent>()

    @Synchronized
    fun append(
        eventType: String,
        actorId: String,
        subjectId: String,
        payload: Map<String, String>,
        occurredAt: Instant,
    ): AuditEvent {
        if (!eventType.matches(Regex("^[A-Z][A-Z0-9_]{2,79}$"))) {
            throw ValidationException("AUDIT_EVENT_INVALID", "Evento de auditoria inválido")
        }
        if (actorId.isBlank() || subjectId.isBlank()) {
            throw ValidationException("AUDIT_IDENTITY_REQUIRED", "Ator e objeto são obrigatórios")
        }
        val sequence = events.size.toLong() + 1L
        val previousHash = events.lastOrNull()?.eventHash ?: "0".repeat(64)
        val eventHash = hash(
            sequence,
            eventType,
            actorId,
            subjectId,
            payload,
            occurredAt,
            previousHash,
        )
        val event = AuditEvent(
            sequence,
            eventType,
            actorId,
            subjectId,
            payload.toSortedMap(),
            occurredAt,
            previousHash,
            eventHash,
        )
        events += event
        return event
    }

    @Synchronized
    fun events(): List<AuditEvent> = events.toList()

    @Synchronized
    fun verify(): Boolean = verify(events)

    companion object {
        fun verify(events: List<AuditEvent>): Boolean {
            var previousHash = "0".repeat(64)
            var expectedSequence = 1L
            for (event in events) {
                if (event.sequence != expectedSequence) return false
                if (event.previousHash != previousHash) return false
                val expectedHash = hash(
                    event.sequence,
                    event.eventType,
                    event.actorId,
                    event.subjectId,
                    event.payload,
                    event.occurredAt,
                    event.previousHash,
                )
                if (event.eventHash != expectedHash) return false
                previousHash = event.eventHash
                expectedSequence += 1L
            }
            return true
        }

        private fun hash(
            sequence: Long,
            eventType: String,
            actorId: String,
            subjectId: String,
            payload: Map<String, String>,
            occurredAt: Instant,
            previousHash: String,
        ): String = sha256(
            canonical(
                listOf(
                    sequence,
                    eventType,
                    actorId,
                    subjectId,
                    payload.toSortedMap(),
                    occurredAt,
                    previousHash,
                ),
            ),
        )
    }
}
